File: python/solution/vehicle.py
from cost_functions import calculate_cost
from collections import namedtuple
from copy import deepcopy
import pdb

Snapshot = namedtuple("Snapshot", ['lane','s','v','a','state'])

class Vehicle(object):
  L = 1
  preferred_buffer = 6 # impacts "keep lane" behavior.

  def __init__(self, lane, s, v, a):
    self.lane = lane
    self.s = s
    self.v = v
    self.a = a
    self.state = "CS"

    # The following are set when configure() is called.
    # All of them are integers.
    self.max_acceleration = None
    self.target_speed = None
    self.lanes_available = None
    self.goal_lane = None
    self.goal_s = None


  def update_state(self, predictions):
    """
    Updates the "state" of the vehicle by assigning one of the
    following values to 'self.state':

    "KL" - Keep Lane
     - The vehicle will attempt to drive its target speed, unless there is
       traffic in front of it, in which case it will slow down.

    "LCL" or "LCR" - Lane Change Left / Right
     - The vehicle will IMMEDIATELY change lanes and then follow longitudinal
       behavior for the "KL" state in the new lane.

    "PLCL" or "PLCR" - Prepare for Lane Change Left / Right
     - The vehicle will find the nearest vehicle in the adjacent lane which is
       BEHIND itself and will adjust speed to try to get behind that vehicle.

    INPUTS
    - predictions
    A dictionary. The keys are ids of other vehicles and the values are arrays
    where each entry corresponds to the vehicle's predicted location at the
    corresponding timestep. The FIRST element in the array gives the vehicle's
    current position. Example (showing a car with id 3 moving at 2 m/s):

    {
      3 : [
        {"s" : 4, "lane": 0},
        {"s" : 6, "lane": 0},
        {"s" : 8, "lane": 0},
        {"s" : 10, "lane": 0},
      ]
    }

    """
    state = self._get_next_state(predictions)
    # pdb.set_trace()
    self.state = state
    # self.state = "KL" # this is an example of how you change state.

  #go through Finite State Machine (FSM) alogrithm
  #to choose the best next possible next
  def _get_next_state(self, predictions):
    states = ["KL", "LCL", "LCR"]

    #remove impossible successor states
    if  self.lane == 0:
      states.remove("LCL")
    if self.lane == (self.lanes_available -1):
      states.remove("LCR")

    costs = []
    #go through each possible state,
    #find rough trajectory to that state and cost for that trajectory
    for state in states:
      predictions_copy = deepcopy(predictions)

      #find trajectory for this state
      trajectory = self._trajectory_for_state(state,predictions_copy)

      #calculate cost for moving to current state using found trajectory
      cost = calculate_cost(self, trajectory, predictions)
      costs.append({"state": state, "cost" : cost})

    #find the state with min cost trajectory
    best = min(costs, key=lambda s: s['cost'])
    return best['state']

  #finds trajectory for passed @param state, for given time horizon
  #and based on passed predictions
  def _trajectory_for_state(self,state,predictions, horizon=5):
    # remember current state
    snapshot = self.snapshot()


    # pretend to be in new proposed state
    self.state = state
    trajectory = [snapshot]
    # pdb.set_trace()
    for i in range(horizon):
      self.restore_state_from_snapshot(snapshot)
      self.state = state
      self.realize_state(predictions)
      assert 0 <= self.lane < self.lanes_available, "{}".format(self.lane)
      self.increment()
      trajectory.append(self.snapshot())

      # need to remove first prediction for each vehicle.
      # pdb.set_trace()
      for v_id, v in predictions.items():
        v.pop(0)

    # restore state from snapshot
    self.restore_state_from_snapshot(snapshot)
    return trajectory

  #stores this (self) Vehicle object's data (lane, s, v, a, state)
  #into a snapshot
  def snapshot(self):
    return Snapshot(self.lane, self.s, self.v, self.a, self.state)

  #Restores this (self) Vehicle object's data (lane, s, v, a, state)
  #from given snapshot
  def restore_state_from_snapshot(self, snapshot):
    s = snapshot
    self.lane = s.lane
    self.s = s.s
    self.v = s.v
    self.a = s.a
    self.state = s.state

  """
  Called by simulator before simulation begins. Sets various
  parameters (target_speed, lanes_available, max_acceleration, goal_lane, goal_s)
  which will impact the ego vehicle.
  """
  def configure(self, road_data):
    """
    Called by simulator before simulation begins. Sets various
    parameters which will impact the ego vehicle.
    """
    self.target_speed = road_data['speed_limit']
    self.lanes_available = road_data["num_lanes"]
    self.max_acceleration = road_data['max_acceleration']
    goal = road_data['goal']
    self.goal_lane = goal[1]
    self.goal_s = goal[0]

  """
  Method override to be used with `print`
  """
  def __repr__(self):
    s = "s:    {}\n".format(self.s)
    s +="lane: {}\n".format(self.lane)
    s +="v:    {}\n".format(self.v)
    s +="a:    {}\n".format(self.a)
    return s

  """
  Updates (s, v) based on delta_t=1
  """
  def increment(self, dt=1):
    self.s += self.v * dt
    self.v += self.a * dt

  """
  Predicts the state (s, v) of vehicle in t seconds (assuming constant acceleration)
  using motion model
  """
  def state_at(self, t):
    """
    Predicts state of vehicle in t seconds (assuming constant acceleration)
    """
    s = self.s + self.v * t + self.a * t * t / 2
    v = self.v + self.a * t
    return self.lane, s, v, self.a

  def collides_with(self, other, at_time=0):
    """
    Simple collision detection.
    """
    l,   s,   v,   a = self.state_at(at_time)
    l_o, s_o, v_o, a_o = other.state_at(at_time)
    return l == l_o and abs(s-s_o) <= L

  def will_collide_with(self, other, timesteps):
    for t in range(timesteps+1):
      if self.collides_with(other, t):
        return True, t
    return False, None

  def realize_state(self, predictions):
    """
    Given a state, realize it by adjusting acceleration and lane.
    Note - lane changes happen instantaneously.
    """
    state = self.state
    if   state == "CS"  : self.realize_constant_speed()
    elif state == "KL"  : self.realize_keep_lane(predictions)
    elif state == "LCL" : self.realize_lane_change(predictions, "L")
    elif state == "LCR" : self.realize_lane_change(predictions, "R")
    elif state == "PLCL": self.realize_prep_lane_change(predictions, "L")
    elif state == "PLCR": self.realize_prep_lane_change(predictions, "R")

  def realize_constant_speed(self):
    self.a = 0

  def _max_accel_for_lane(self, predictions, lane, s):
    delta_v_til_target = self.target_speed - self.v
    max_acc = min(self.max_acceleration, delta_v_til_target)
    in_front = [v for (v_id, v) in predictions.items() if v[0]['lane'] == lane and v[0]['s'] > s ]
    if len(in_front) > 0:
      leading = min(in_front, key=lambda v: v[0]['s'] - s)
      next_pos = leading[1]['s']
      my_next = s + self.v
      separation_next = next_pos - my_next
      available_room = separation_next - self.preferred_buffer
      max_acc = min(max_acc, available_room)
    return max_acc

  def realize_keep_lane(self, predictions):
    self.a = self._max_accel_for_lane(predictions, self.lane, self.s)

  def realize_lane_change(self, predictions, direction):
    delta = -1
    if direction == "R": delta = 1
    self.lane += delta
    self.a = self._max_accel_for_lane(predictions, self.lane, self.s)

  def realize_prep_lane_change(self, predictions, direction):
    delta = -1
    if direction == "L": delta = 1
    lane = self.lane + delta
    ids_and_vehicles = [(v_id, v) for (v_id, v) in predictions.items() if v[0]['lane'] == lane and v[0]['s'] <= self.s]
    if len(ids_and_vehicles) > 0:
      vehicles = [v[1] for v in ids_and_vehicles]

      nearest_behind = max(ids_and_vehicles, key=lambda v: v[1][0]['s'])

      print("nearest behind : {}").format(nearest_behind)
      nearest_behind = nearest_behind[1]
      target_vel = nearest_behind[1]['s'] - nearest_behind[0]['s']
      delta_v = self.v - target_vel
      delta_s = self.s - nearest_behind[0]['s']
      if delta_v != 0:
        print("delta_v {}").format(delta_v)
        print("delta_s {}").format(delta_s)
        time = -2 * delta_s / delta_v
        if time == 0:
          a = self.a
        else:
          a = delta_v / time
        print("raw a is {}").format(a)
        if a > self.max_acceleration: a = self.max_acceleration
        if a < -self.max_acceleration: a = -self.max_acceleration
        self.a = a
        print("time : {}").format(time)
        print("a: {}").format(self.a)
      else :
        min_acc = max(-self.max_acceleration, -delta_s)
        self.a = min_acc

  """
  Predicts state (s, v, a) of vehicle for given number of timesteps (horizon, by default 10)
  in seconds (assuming constant acceleration) and appends each predicted state
  to predictions dictionary
  """
  def generate_predictions(self, horizon=10):
    predictions = []
    for i in range(horizon):
      lane, s, v, a = self.state_at(i)
      predictions.append({'s':s, 'lane': lane})
    return predictions

File: src/vehicle.cpp
/*
 * Vehicle.cpp
 *
 *  Created on: Jul 29, 2017
 *      Author: ramiz
 */

#include <cassert>
#include <iostream>
#include <algorithm>
#include "vehicle.h"
#include <math.h>
#include <map>
#include <string>
#include <iterator>
#include "cost_functions.h"

/**
 * Initializes Vehicle
 */
Vehicle::Vehicle(int lane, int s, int v, int a) {

  this->lane = lane;
  this->s = s;
  this->v = v;
  this->a = a;
  state = "CS";
  max_acceleration = -1;

}

Vehicle::~Vehicle() {
}

// TODO - Implement this method.
void Vehicle::update_state(map<int, vector<vector<int> > > predictions) {
  /*
   Updates the "state" of the vehicle by assigning one of the
   following values to 'self.state':

   "KL" - Keep Lane
   - The vehicle will attempt to drive its target speed, unless there is
   traffic in front of it, in which case it will slow down.

   "LCL" or "LCR" - Lane Change Left / Right
   - The vehicle will IMMEDIATELY change lanes and then follow longitudinal
   behavior for the "KL" state in the new lane.

   "PLCL" or "PLCR" - Prepare for Lane Change Left / Right
   - The vehicle will find the nearest vehicle in the adjacent lane which is
   BEHIND itself and will adjust speed to try to get behind that vehicle.

   INPUTS
   - predictions
   A dictionary. The keys are ids of other vehicles and the values are arrays
   where each entry corresponds to the vehicle's predicted location at the
   corresponding timestep. The FIRST element in the array gives the vehicle's
   current position. Example (showing a car with id 3 moving at 2 m/s):

   {
   3 : [
   {"s" : 4, "lane": 0},
   {"s" : 6, "lane": 0},
   {"s" : 8, "lane": 0},
   {"s" : 10, "lane": 0},
   ]
   }

   */
  state = get_state(predictions);
}

/**
 * @returns new state for vehicle based on predictions of other vehicles given
 */
string Vehicle::get_state(const map<int, vector<vector<int> > > &predictions) {
  vector<string> possible_states = {"KL", "LCL", "LCR"};

  //check if we are in left-most-lane
  if (this->lane == 0) {
    //we are in left-most lane so Lane Change Left (LCL)
    //is not possible, remove it from possible states
    possible_states.erase(possible_states.begin() + 1);
  }

  //check if we are in right most lane
  if (this->lane == (this->lanes_available - 1)) {
    //we are in right-most lane so Lane Change Left (LCR)
    //is not possible, remove it from possible states
    possible_states.erase(possible_states.begin() + 2);
  }

  vector<double> costs;
  CostFunctions cost_functions;
  for(int i = 0; i < possible_states.size(); ++i) {
    //find a rough trajectory to reach this state
    vector<Snapshot> trajectory = trajectory_for_state(possible_states[i], predictions);
    //find cost for found trajectory
    double cost = cost_functions.calculate_cost(*this, predictions, trajectory);
    costs.push_back(cost);
  }

  //find index of min cost which is also index of corresponding state
  int min_cost_state_index = std::distance(costs.begin(), std::min_element(costs.begin(), costs.end()));

  //return min cost state
  return possible_states[min_cost_state_index];
}

/**
 * @param state, state for which trajectory to calculate
 * @param predictions, map of predictions of other vehicle's predicted trajectories
 * @returns Returns trajectory to for reaching given state
 */
vector<Snapshot> Vehicle::trajectory_for_state(const string &state,
                                               const map<int, vector<vector<int> > > &predictions,
                                               int horizon) {
  //make a copy of predictions are we are going to change it
  map<int, vector<vector<int> > > predictions_copy = predictions;

  //make a vector to hold trajectory
  vector<Snapshot> trajectory;
  //take snapshot of current vehicle state
  Snapshot current_state_snapshot = take_current_state_snapshot();

  //add current state snapshot to trajectory
  trajectory.push_back(current_state_snapshot);

  //for prediction time horizon
  for (int i = 0; i < horizon; ++i) {
    //reset vehicle state to its original state (current_state_snapshot)
    restore_state_from_snapshot(current_state_snapshot);

    //pretend to be in new (passed as param) state
    this->state = state;

    //update vehicle to passed state by changing (lane, a) based on given state and
    //predictions of other vehicles
    realize_state(predictions_copy);

    //validate lane number
    assert(0 <= this->lane && this->lane < this->lanes_available);

    //update current vehicle (s, v) for delta_t=1, assuming constant acceleration
    increment();

    //add current state snapshot to trajectory
    trajectory.push_back(take_current_state_snapshot());

    //As one timestep has been finished so we
    //need to remove first prediction for each vehicle
    //so that all vehicles predictions start from the next timestep
    remove_first_prediction_for_each(predictions_copy);
  }

  //reset state to original state (current_state_snapshot)
  restore_state_from_snapshot(current_state_snapshot);

  return trajectory;
}

void Vehicle::remove_first_prediction_for_each(map<int, vector<vector<int> > > &predictions) {
  map<int, vector<vector<int> > >::iterator iter = predictions.begin();

  while (iter != predictions.end()) {
    //map value (second element in iterator) is vector with each vector value a vector: [s, lane]
    //take a reference to this second vector object so that we can update it
    vector<vector<int> > &v_preds = iter->second;

    //remove first vector of this prediction
    v_preds.erase(v_preds.begin());

    //move to next vehicle predictions
    iter++;
  }
}

/**
 * Takes snapshot of current state of vehicle and saves it into snapshot object
 * @return  Snapshot object containing current state of vehicle
 */
Snapshot Vehicle::take_current_state_snapshot() {
  return Snapshot(this->lane,
      this->s,
      this->v,
      this->a,
      this->state);
}
/**
 * Sets current state of vehicle to that of saved in passed snapshot
 * @param snapshot  snapshot containing state of vehicle to restore from
 */
void Vehicle::restore_state_from_snapshot(const Snapshot& snapshot) {
  this->lane = snapshot.lane;
  this->s = snapshot.s;
  this->v = snapshot.v;
  this->a = snapshot.a;
  this->state = snapshot.state;
}

void Vehicle::configure(vector<int> road_data) {
  /*
   Called by simulator before simulation begins. Sets various
   parameters which will impact the ego vehicle.
   vector<int> ego_config = { SPEED_LIMIT, num_lanes, goal_s, goal_lane,
        MAX_ACCEL };
   */
  target_speed = road_data[0];
  lanes_available = road_data[1];
  goal_s = road_data[2];
  goal_lane = road_data[3];
  max_acceleration = road_data[4];
}

string Vehicle::display() {

  ostringstream oss;

  oss << "s:    " << this->s << "\n";
  oss << "lane: " << this->lane << "\n";
  oss << "v:    " << this->v << "\n";
  oss << "a:    " << this->a << "\n";

  return oss.str();
}

void Vehicle::increment(int dt) {

  this->s += this->v * dt;
  this->v += this->a * dt;
}

vector<int> Vehicle::state_at(int t) {

  /*
   Predicts state of vehicle in t seconds (assuming constant acceleration)
   */
  int s = this->s + this->v * t + this->a * t * t / 2;
  int v = this->v + this->a * t;
  return {this->lane, s, v, this->a};
}

bool Vehicle::collides_with(Vehicle other, int at_time) {

  /*
   Simple collision detection.
   */
  vector<int> check1 = state_at(at_time);
  vector<int> check2 = other.state_at(at_time);
  return (check1[0] == check2[0]) && (abs(check1[1] - check2[1]) <= L);
}

Vehicle::collider Vehicle::will_collide_with(Vehicle other, int timesteps) {

  Vehicle::collider collider_temp;
  collider_temp.collision = false;
  collider_temp.time = -1;

  for (int t = 0; t < timesteps + 1; t++) {
    if (collides_with(other, t)) {
      collider_temp.collision = true;
      collider_temp.time = t;
      return collider_temp;
    }
  }

  return collider_temp;
}

void Vehicle::realize_state(map<int, vector<vector<int> > > predictions) {

  /*
   Given a state, realize it by adjusting acceleration and lane.
   Note - lane changes happen instantaneously.
   */
  string state = this->state;
  if (state.compare("CS") == 0) {
    realize_constant_speed();
  } else if (state.compare("KL") == 0) {
    realize_keep_lane(predictions);
  } else if (state.compare("LCL") == 0) {
    realize_lane_change(predictions, "L");
  } else if (state.compare("LCR") == 0) {
    realize_lane_change(predictions, "R");
  } else if (state.compare("PLCL") == 0) {
    realize_prep_lane_change(predictions, "L");
  } else if (state.compare("PLCR") == 0) {
    realize_prep_lane_change(predictions, "R");
  }

}

void Vehicle::realize_constant_speed() {
  a = 0;
}

int Vehicle::_max_accel_for_lane(map<int, vector<vector<int> > > predictions,
                                 int lane, int s) {

  int delta_v_til_target = target_speed - v;
  int max_acc = min(max_acceleration, delta_v_til_target);

  map<int, vector<vector<int> > >::iterator it = predictions.begin();
  vector<vector<vector<int> > > in_front;
  while (it != predictions.end()) {

    int v_id = it->first;

    vector<vector<int> > v = it->second;

    if ((v[0][0] == lane) && (v[0][1] > s)) {
      in_front.push_back(v);

    }
    it++;
  }

  if (in_front.size() > 0) {
    int min_s = 1000;
    vector<vector<int>> leading = { };
    for (int i = 0; i < in_front.size(); i++) {
      if ((in_front[i][0][1] - s) < min_s) {
        min_s = (in_front[i][0][1] - s);
        leading = in_front[i];
      }
    }

    int next_pos = leading[1][1];
    int my_next = s + this->v;
    int separation_next = next_pos - my_next;
    int available_room = separation_next - preferred_buffer;
    max_acc = min(max_acc, available_room);
  }

  return max_acc;

}

void Vehicle::realize_keep_lane(map<int, vector<vector<int> > > predictions) {
  this->a = _max_accel_for_lane(predictions, this->lane, this->s);
}

void Vehicle::realize_lane_change(map<int, vector<vector<int> > > predictions,
                                  string direction) {
  int delta = -1;
  if (direction.compare("L") == 0) {
    delta = 1;
  }
  this->lane += delta;
  int lane = this->lane;
  int s = this->s;
  this->a = _max_accel_for_lane(predictions, lane, s);
}

void Vehicle::realize_prep_lane_change(
    map<int, vector<vector<int> > > predictions, string direction) {
  int delta = -1;
  if (direction.compare("L") == 0) {
    delta = 1;
  }
  int lane = this->lane + delta;

  map<int, vector<vector<int> > >::iterator it = predictions.begin();
  vector<vector<vector<int> > > at_behind;
  while (it != predictions.end()) {
    int v_id = it->first;
    vector<vector<int> > v = it->second;

    if ((v[0][0] == lane) && (v[0][1] <= this->s)) {
      at_behind.push_back(v);

    }
    it++;
  }
  if (at_behind.size() > 0) {

    int max_s = -1000;
    vector<vector<int> > nearest_behind = { };
    for (int i = 0; i < at_behind.size(); i++) {
      if ((at_behind[i][0][1]) > max_s) {
        max_s = at_behind[i][0][1];
        nearest_behind = at_behind[i];
      }
    }
    int target_vel = nearest_behind[1][1] - nearest_behind[0][1];
    int delta_v = this->v - target_vel;
    int delta_s = this->s - nearest_behind[0][1];
    if (delta_v != 0) {

      int time = -2 * delta_s / delta_v;
      int a;
      if (time == 0) {
        a = this->a;
      } else {
        a = delta_v / time;
      }
      if (a > this->max_acceleration) {
        a = this->max_acceleration;
      }
      if (a < -this->max_acceleration) {
        a = -this->max_acceleration;
      }
      this->a = a;
    } else {
      int my_min_acc = max(-this->max_acceleration, -delta_s);
      this->a = my_min_acc;
    }

  }

}

vector<vector<int> > Vehicle::generate_predictions(int horizon = 10) {

  vector<vector<int> > predictions;
  for (int i = 0; i < horizon; i++) {
    vector<int> check1 = state_at(i);
    vector<int> lane_s = { check1[0], check1[1] };
    predictions.push_back(lane_s);
  }
  return predictions;

}

